network/ocsp.go

```diff
@@ -0,0 +1,334 @@
+
+		if v.Config.fromCert == ocspFromCertPrefer || v.Config.fromCert == ocspFromCertTrust {
+			log.Debugf("OCSP: appending server %s, from config", serverToCheck.url)
+			serversToCheck = append(serversToCheck, serverToCheck)
+		} else {
+			log.Debugf("OCSP: prepending server %s, from config", serverToCheck.url)
+			serversToCheck = append([]ocspServerToCheck{serverToCheck}, serversToCheck...)
+		}
+	}
+
+	queriedOCSPs := make(map[string]struct{})
+
+	confirms := 0
+
+	for _, serverToCheck := range serversToCheck {
+		log.Debugf("OCSP: Trying server %s", serverToCheck.url)
+
+		if _, ok := queriedOCSPs[serverToCheck.url]; ok {
+			log.Debugln("OCSP: Skipping, already queried")
+			continue
+		}
+
+		response, err := v.Client.Query(cert.Issuer.CommonName, cert, issuer, serverToCheck.url)
+		if err != nil {
+			log.WithError(err).WithField("uri", serverToCheck.url).Warnln("Cannot query OCSP server")
+
+			if v.Config.required == ocspRequiredAll {
+				return ErrOCSPRequiredAllButGotError
+			}
+
+			continue
+		}
+
+		switch response.Status {
+		case ocsp.Good:
+			confirms++
+
+			if serverToCheck.fromCert {
+				log.Debugln("OCSP: confirmed by server from certificate")
+			} else {
+				log.Debugln("OCSP: confirmed by server from config")
+			}
+
+			if v.Config.required != ocspRequiredAll {
+				// One confirmation is enough if we don't require all OCSP servers to confirm the certificate validity
+				break
+			}
+
+			if serverToCheck.fromCert && v.Config.fromCert == ocspFromCertTrust {
+				// If this OCSP server came from certificate and `--tls_ocsp_from_cert=trust`, don't perform further checks
+				break
+			}
+		case ocsp.Revoked:
+			// If any OCSP server replies with "certificate was revoked", return error immediately
+			log.WithField("serial", cert.SerialNumber).WithField("revoked_at", response.RevokedAt).Warnln("OCSP: Certificate was revoked")
+			return ErrCertWasRevoked
+		case ocsp.Unknown:
+			// Treat "Unknown" response as error if tls_ocsp_required is "yes" or "all"
+			if v.Config.required != ocspRequiredNo {
+				log.WithField("uri", serverToCheck.url).WithField("serial", cert.SerialNumber).Warnln("OCSP server doesn't know about certificate")
+				return ErrOCSPUnknownCertificate
+			}
+		}
+
+		queriedOCSPs[serverToCheck.url] = struct{}{}
+	}
+
+	if len(serversToCheck) > 0 && confirms == 0 {
+		return ErrOCSPNoConfirms
+	}
+	return nil
+}
+
+// Verify ensures certificate is not revoked by querying configured OCSP servers
+func (v DefaultOCSPVerifier) Verify(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+	for _, chain := range verifiedChains {
+		if len(chain) == 1 {
+			// This one cert[0] must be trusted since it was allowed by more basic verifying routines.
+			// If we are at this point, we have nothing to do, and no CA means no OCSP.
+			log.Debugln("OCSP: Certificate chain contains one certificate, nothing to do")
+			return nil
+		}
```

Comment:
I'm sorry for joining this discussion a bit late. 

It seems to me that (following best practices) chain should contain at least two certificates - one is self-signed CA certificate and the second - is the certificate (signed by this CA) issued to the entity that we verify. We can trust the entity's certificate only if we trust to this CA. The situation when only one self-signed certificate (1 certificate in chain) is provided while handshake, is technically correct - it means that we trust to this entity directly (we can definitely run TLS handshake protocol). But such authentication is quite rare case. It means that we already have knowledge of entity's public key (e.g. during some prior installation phase). Alternatively, such authentication has no sense at al. In context of CRL/OCSP it means that we definitely can skip CRL/OCSP verification, since such entity is a Certificate Authority itself. And it will unlikely decide to revoke its own credential to TLS communication from itself.

Reply:
Thanks @storojs72 for explaining this. Now that you've framed it like this:

> In context of CRL/OCSP it means that we definitely can skip CRL/OCSP verification, since such entity is a Certificate Authority itself. And it will unlikely decide to revoke its own credential to TLS communication from itself.

I believe that it would be okay to leave a small comment describing this and a log line to inform the users that Acra has skipped redundant CRL/OCSP verification for an ultimately trusted certificate.

Reply:
Left the same comment, changed logging code to
```go
log.WithField("serial", chain[0].SerialNumber).WithField("subject", chain[0].Subject).
	Infoln("OCSP: Certificate chain consists of one already trusted certificate, nothing to do")
```
done the same thing to CRL, thought it would be useful to see in logs which certificate is used in this weird way

Reply:
After last discussion, made this log message a warning, removed `subject` field and changed the message text. Everything else seems to be good enough.